File: src/gestorAplicacion/Veterinaria/Medico.java
package gestorAplicacion.Veterinaria;
import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

import gestorAplicacion.Clientes.Cliente;
import gestorAplicacion.Clientes.Comunicacion;

public class Medico extends Persona implements Comunicacion, Serializable{
    /**
	 * 
	 */
	private static final long serialVersionUID = 3L;
	public  Map<String, Turno[]> agenda = new HashMap<>();
	public static Map<String, Medico> mapaMedico = new HashMap<>();
    public tipoMedico tipoMed;//general o especialista

    public Medico(String nombre, String cedula, String telefono, tipoMedico tipoMed) {
        super(nombre, cedula, telefono);
        this.tipoMed = tipoMed;
        mapaMedico.put(cedula, this);
    }
    
    
    public tipoMedico getTipoMed(){
        return tipoMed;
    }
    
    public void setTipoMed(tipoMedico tipoMed) {
        this.tipoMed = tipoMed;
    } 
    
    public void crearFecha(String fecha) { // valida si la fecha ingresada ya esta contemplada en la agenda del doctor, si no lo esta crea un arreglo de 24 posiciones con turno vacios y lo ingresa a la agenda del doctor
    	
    	if(!this.agenda.containsKey(fecha)) {
    		Turno[] turnos = new Turno[24];
    		for(int i = 1; i<25; i++) {
    			turnos[i-1] = new Turno((i-1),fecha,null,null,true,this);
    		}
    		this.agenda.put(fecha, turnos);
		}
		
	}
    public static String obtenerMedicos(tipoMedico tipoMed) { //recibe una constante del enum y devuelve los medicos cuyo atributo tipoMed concuerda con la constante
    	String cadena = "";
    	
    	for (Map.Entry<String, Medico> entry : Medico.mapaMedico.entrySet()) {
    		if(entry.getValue().getTipoMed().equals(tipoMed)){
    			cadena = cadena + (entry.getValue().getNombre()+" - Cedula: "+ entry.getValue().getCedula() +"\n");// acumula en un string los datos de todos los medicos que cumplen con la condicion
    		}
	    }
    	return cadena;
    }    
    public String obtenerTurnosDisponibles(String fecha) { //consulta la agenda del doctor en la fecha ingresada y devuelve un string con la acumulacion de todo los turnos cuyo atributo disponible es true
    	String cadena = "";
    	for(int i = 0; i<24; i++) {
			if(this.agenda.get(fecha)[i].isDisponibilidad()) {
				if(this.agenda.get(fecha)[i].getHoraInicio()<13) { //acumula un string diferente en cadena dependiendo de si el turno es AM o PM
					cadena = cadena + "Turno "+(i+1)+": "+this.agenda.get(fecha)[i].getHoraInicio()+":00 AM\n";
				}else {
					cadena = cadena + "Turno "+(i+1)+": "+this.agenda.get(fecha)[i].getHoraInicio()+":00 PM\n";
				}
			}	
		}
    	return cadena;
    	
    }
    public void asignarTurno(String fecha, int turno, String cedulaCliente, int mascota) { // recibe la informacion necesaria 
    	this.agenda.get(fecha)[turno].setDisponibilidad(false); //el objeto turno ya esta instanciado en la agenda del medico, solo se usan los set para darle valor a sus atributos
		this.agenda.get(fecha)[turno].setCliente(Cliente.mapaClientes.get(cedulaCliente));
		this.agenda.get(fecha)[turno].setMascota(Cliente.mascotas.get(cedulaCliente).get(mascota));
		Cliente.mapaClientes.get(cedulaCliente).registrarHora(turno+1); //guarda la hora del turno en un hassmap que se usara despues para recomendar horarios
		if(Cliente.mapaClientes.get(cedulaCliente).sumaRegistros()>3) { //si el numero de registro en el hasmap es mayor que 3, establece el cliente como un cliente frecuente
			Cliente.mapaClientes.get(cedulaCliente).setFrecuente(true);
		}
		Cliente.mapaClientes.get(cedulaCliente).turnosPendientes.add(this.agenda.get(fecha)[turno]);
    }
    public static boolean validarCedula(String cedula) { //valida si la cedula se encuentra registrada en el sistema
		return Medico.mapaMedico.containsKey(cedula);
	}
    public String toString() {
		return "nombre: " + nombre + "\n" + "cedula: " + cedula + "\n" + "telefono: " + telefono;
	}

 
	@Override
	public String saludar() {  //metodo abstracto que se hereda de personal
		return "Hola soy medico";
	}
	
	public String presentarse() {
		return "El medico " + this.nombre + " fue registrado";
	}
	@Override
	public String llamar() {//metodo que se hereda de comunicacion
		return "Llamo desde mi telefono " + this.telefono;
	}

}

File: Python/gestorAplicacion/Veterinaria/Medico.py
from gestorAplicacion.Veterinaria.Persona import Persona
from gestorAplicacion.Veterinaria.Turno import Turno
from gestorAplicacion.Clientes import Cliente
class Medico(Persona):
    mapaMedico= dict()
    
    def __init__(self, nombre, cedula, telefono, cargo):
        super().__init__(nombre, cedula, telefono)
        self.cargo = cargo
        self.agenda = dict()
    
    def getTipoMedico(self):
        return self.cargo
    
    def setTipoMedico(self, cargo):
        self.cargo = cargo
        
    def crearFecha(self, fecha):
        if(fecha not in self.agenda):
            turnos = []
            
            for i in range(24):
                turnos.append(Turno(i, fecha, None, None, True, self))
            self.agenda[fecha] = turnos
            
    def obtenerTurnosDisponibles(self, fecha):
        lista = []
        for i in range(24):
            if(self.agenda[fecha][i].isDisponibilidad()):
                if(self.agenda[fecha][i].getHoraInicio()<13):
                    lista.append("Turno "+str(i+1)+": "+str(self.agenda[fecha][i].getHoraInicio())+":00 AM")
                else:
                    lista.append("Turno "+str(i+1)+": "+str(self.agenda[fecha][i].getHoraInicio())+":00 PM")
        if(len(lista)==0):
            return 0
        else:
            return lista
        
        
    def asignarTurno(self, fecha, turno, cedulaCliente, mascota): # recibe la informacion necesaria 
        self.agenda[fecha][turno].setDisponibilidad(False); #el objeto turno ya esta instanciado en la agenda del medico, solo se usan los set para darle valor a sus atributos
        self.agenda[fecha][turno].setCliente(Cliente.Cliente.mapaClientes[cedulaCliente])
        self.agenda[fecha][turno].setMascota(Cliente.Cliente.mascotas[cedulaCliente][mascota])
        Cliente.Cliente.mapaClientes.get(cedulaCliente).registrarHora(turno) #guarda la hora del turno en un hassmap que se usara despues para recomendar horarios
        
        if(Cliente.Cliente.mapaClientes.get(cedulaCliente).sumaRegistros()>3): #si el numero de registro en el hasmap es mayor que 3, establece el cliente como un cliente frecuente
            Cliente.Cliente.mapaClientes[cedulaCliente].setFrecuente(True)
            
        Cliente.Cliente.mapaClientes[cedulaCliente].turnosPendientes.append(self.agenda[fecha][turno])
        
    
        
    @classmethod
    def obtenerNombresMedicos(cls, tipo):
        lista=[]
        for cedula in Medico.mapaMedico:
            if(Medico.mapaMedico[cedula].getTipoMedico()==tipo):
                lista.append(Medico.mapaMedico[cedula].getNombre())
        if(len(lista)==0):
            return 0
        else:
            return lista
    
    @classmethod
    def obtenerCedulasMedicos(cls, tipo):
        lista=[]
        for cedula in Medico.mapaMedico:
            if(Medico.mapaMedico[cedula].getTipoMedico()==tipo):
                lista.append(Medico.mapaMedico[cedula].getCedula())
        if(len(lista)==0):
            return 0
        else:
            return lista